File: go/specialized/primeSieve.go
package specialized

import (
	"fmt"
	"os"
	"strconv"
)

func primeSieve(num int) []bool {
	// make defaults to 0/false so I'm gonna flip the semantics and say false is true and true is false
	var primes = make([]bool, num)
	primes[0] = true
	primes[1] = true
	for i := 2; i < num; i++ {
		if !primes[i] {
			var j int = i
			var current int = j * i
			for current < num {
				primes[current] = true
				j++
				current = j * i
			}
		}

	}
	return primes
}

func main4() {
	var max int = 42
	if len(os.Args) > 1 { // program name is always the first argument
		max, _ = strconv.Atoi(os.Args[1])
	}

	var temp []bool = primeSieve(max)
	for i := 2; i < len(temp); i++ {
		if !temp[i] {
			fmt.Println(i)
		}
	}
}

File: java/specialized/primeSieve.java
package specialized;
public class primeSieve {
    public static boolean[] sieve(int num) {
        boolean[] primes = new boolean[num];
        // default instantiation is false, flipping semantics
        primes[0] = true;
        primes[1] = true;
        for (int i = 2; i < num; i++) {
            if (!primes[i]) {
                int j = i;
                while (j * i < num && j * i > 0) { // check for overflow
                    primes[j*i] = true;
                    j++;
                }
            }
        }
        return primes;
    }
    
    public static void main(String[] args) throws Exception {
        int max = 42;
        if (args.length > 0){
            max = Integer.parseInt(args[0]);
        }
        boolean[] temp = sieve(max);
        for (int i = 2; i < max; i++) {
            if (!temp[i]) {
                System.out.println(i);
            }
        }
    }
}
